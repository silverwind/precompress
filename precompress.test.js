import {deleteSync} from "del";
import {execa} from "execa";
import {temporaryDirectory} from "tempy";
import {fileURLToPath} from "node:url";
import {writeFileSync, readFileSync, mkdirSync, statSync} from "node:fs";
import {join} from "node:path";
import fastGlob from "fast-glob";

const testDir = temporaryDirectory();
const script = fileURLToPath(new URL("bin/precompress.js", import.meta.url));

beforeEach(() => {
  deleteSync(join(testDir, "*"), {force: true});
  writeFileSync(join(testDir, "index.html"), (new Array(1e4)).join("index"));
  writeFileSync(join(testDir, "already.gz"), (new Array(1e4)).join("index"));
  writeFileSync(join(testDir, "image.png"), (new Array(1e4)).join("image"));
  mkdirSync(join(testDir, "src"));
  writeFileSync(join(testDir, "index.js"), (new Array(1e4)).join("index"));
  writeFileSync(join(testDir, "index.css"), (new Array(1e4)).join("index"));
});

afterAll(() => {
  deleteSync(testDir, {force: true});
});

async function run(args) {
  const argsArr = [".", ...args.split(/\s+/).map(s => s.trim()).filter(Boolean)];
  return execa(script, argsArr, {cwd: testDir});
}

function makeTest(args) {
  return async () => {
    await run(args);
    const paths = fastGlob.sync(`**/*`, {cwd: testDir}).sort()
      .filter(p => !statSync(join(testDir, p)).isDirectory());
    expect(paths).toMatchSnapshot();
  };
}

test("version", async () => {
  const {version: expected} = JSON.parse(readFileSync(new URL("package.json", import.meta.url), "utf8"));
  const {stdout, exitCode} = await execa("node", [script, "-v"]);
  expect(stdout).toEqual(expected);
  expect(exitCode).toEqual(0);
});

test("simple", makeTest(""));
test("include", makeTest("-i html,foo"));
test("include #2", makeTest("-i HTML"));
test("exclude", makeTest("-e png"));
test("exclude #2", makeTest("-e png -e png,png"));
test("exclude #3", makeTest("-e html"));
test("exclude #4", makeTest("-e ''"));
test("mtime", makeTest("-m"));
test("outdir", makeTest(`-o ${testDir}/dist`)); // TODO: path wrong
test("outdir,extensionless", makeTest(`-o ${testDir}/dist -t gz -E`)); // TODO: include already compressed when extensionless

test("no matching files", async () => {
  await expect(run("-e png,html,js,css")).rejects.toThrow();
  await expect(run("-i HTML -S")).rejects.toThrow();
});
