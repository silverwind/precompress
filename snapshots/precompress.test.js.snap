// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exclude #2 1`] = `
[
  "already.gz",
  "image.png",
  "index.css",
  "index.css.br",
  "index.css.gz",
  "index.html",
  "index.html.br",
  "index.html.gz",
  "index.js",
  "index.js.br",
  "index.js.gz",
]
`;

exports[`exclude #3 1`] = `
[
  "already.gz",
  "image.png",
  "image.png.br",
  "image.png.gz",
  "index.css",
  "index.css.br",
  "index.css.gz",
  "index.html",
  "index.js",
  "index.js.br",
  "index.js.gz",
]
`;

exports[`exclude #4 1`] = `
[
  "already.gz",
  "image.png",
  "image.png.br",
  "image.png.gz",
  "index.css",
  "index.css.br",
  "index.css.gz",
  "index.html",
  "index.html.br",
  "index.html.gz",
  "index.js",
  "index.js.br",
  "index.js.gz",
]
`;

exports[`exclude 1`] = `
[
  "already.gz",
  "image.png",
  "index.css",
  "index.css.br",
  "index.css.gz",
  "index.html",
  "index.html.br",
  "index.html.gz",
  "index.js",
  "index.js.br",
  "index.js.gz",
]
`;

exports[`include #2 1`] = `
[
  "already.gz",
  "image.png",
  "index.css",
  "index.html",
  "index.html.br",
  "index.html.gz",
  "index.js",
]
`;

exports[`include 1`] = `
[
  "already.gz",
  "image.png",
  "index.css",
  "index.html",
  "index.html.br",
  "index.html.gz",
  "index.js",
]
`;

exports[`mtime 1`] = `
[
  "already.gz",
  "image.png",
  "image.png.br",
  "image.png.gz",
  "index.css",
  "index.css.br",
  "index.css.gz",
  "index.html",
  "index.html.br",
  "index.html.gz",
  "index.js",
  "index.js.br",
  "index.js.gz",
]
`;

exports[`outdir 1`] = `
[
  "already.gz",
  "dist/image.png.br",
  "dist/image.png.gz",
  "dist/index.css.br",
  "dist/index.css.gz",
  "dist/index.html.br",
  "dist/index.html.gz",
  "dist/index.js.br",
  "dist/index.js.gz",
  "image.png",
  "index.css",
  "index.html",
  "index.js",
]
`;

exports[`outdir,extensionless 1`] = `
[
  "already.gz",
  "dist/image.png",
  "dist/index.css",
  "dist/index.html",
  "dist/index.js",
  "image.png",
  "index.css",
  "index.html",
  "index.js",
]
`;

exports[`simple 1`] = `
[
  "already.gz",
  "image.png",
  "image.png.br",
  "image.png.gz",
  "index.css",
  "index.css.br",
  "index.css.gz",
  "index.html",
  "index.html.br",
  "index.html.gz",
  "index.js",
  "index.js.br",
  "index.js.gz",
]
`;
